README.md
# unqlite

A high-level UnQLite database engine wrapper.

[![travis-badge][]][travis] [![release-badge][]][cargo] [![downloads]][cargo]
[![docs-badge][]][docs] [![license-badge][]][cargo]

NOTE: Some of the documents is stolen from [UnQLite Official Website][unqlite].

## What is UnQLite?

>
UnQLite is a software library which implements a *self-contained*, *serverless*,
zero-configuration, transactional NoSQL database engine. UnQLite is a document store database
similar to [MongoDB], [Redis], [CouchDB] etc. as well a standard Key/Value store similar to
[BerkeleyDB], [LevelDB], etc.
>
UnQLite is an embedded NoSQL (Key/Value store and Document-store) database engine. Unlike most
other NoSQL databases, UnQLite does not have a separate server process. UnQLite reads and
writes directly to ordinary disk files. A complete database with multiple collections, is
contained in **a single disk file**. The database file format is cross-platform, you can freely
copy a database between 32-bit and 64-bit systems or between big-endian and little-endian
architectures.

## Port to Rust

This crate is high-level UnQLite database wrapper for Rust. A low-level bindings wrapper
is avaliable as a seperated crate: [unqlite-sys](https://crates.io/crates/unqlite-sys).

## Usage

You can start with `UnQLite` constructors:

```rust
extern crate unqlite;

use unqlite::{UnQLite, Config, KV, Cursor};

fn main() {
    // The database memory is not handled by Rust, and the database is on-disk,
    // so `mut` is not neccessary.
    let unqlite = UnQLite::create_temp();
    // Use any type that can use as `[u8]`
    unqlite.kv_store("key", "a long length value").unwrap();
    unqlite.kv_store("abc", [1,2,3]).unwrap();

    let mut entry = unqlite.first();
    // Iterate records
    loop {
        if entry.is_none() { break; }

        let record = entry.expect("valid entry");
        let (key, value) = record.key_value();
        println!("* Go through {:?} --> {:?}", key, value);

        if value.len() > 10 {
            println!("** Delete key {:?} by value length", key);
            entry = record.delete();
        } else {
            entry = record.next();
        }
    }
    //panic!("for test");
}
```

## Contributors

- **@bemyak**
- **@chritchens**
- **@wolandr**
- **@timlyo**
- **@dariusc93**

[unqlite]: https://unqlite.org/index.html
[travis-badge]: https://img.shields.io/travis/zitsen/unqlite.rs.svg?style=flat-square
[travis]: https://travis-ci.org/zitsen/unqlite.rs
[release-badge]: https://img.shields.io/crates/v/unqlite.svg?style=flat-square
[downloads]: https://img.shields.io/crates/d/unqlite.svg?style=flat-square
[cargo]: https://crates.io/crates/unqlite
[docs-badge]: https://img.shields.io/badge/API-docs-blue.svg?style=flat-square
[docs]: https://zitsen.github.io/unqlite.rs
[license-badge]: https://img.shields.io/crates/l/unqlite.svg?style=flat-square


Changelog.md
<a name="v1.5.0"></a>
## v1.5.0 (2020-10-16)


#### Features

*   remove deprecated functions and update to 2018 edition, thanks @dariusc93 ([441c36b0](441c36b0))



<a name="v1.4.3"></a>
## v1.4.3 (2019-06-27)


#### Bug Fixes

*   use dyn for bare trait object warning in `Error` ([db9120f9](db9120f9))



<a name="v1.4.2"></a>
## v1.4.2 (2019-05-21)


#### Bug Fixes

* **tests:**  fix locked error in some multi-threading cases ([b40cc5ac](b40cc5ac))



<a name="v1.4.1"></a>
### v1.4.1 (2018-05-16)


#### Features

* **UnQLiteVm:**  Add Clone trait to UnQLiteVm ([53b977a4](53b977a4))



<a name="v1.4.0"></a>
## v1.4.0 (2018-01-26)


#### Bug Fixes

* **util:**  warn use to not use `load_mmaped_file` ([ded1052e](ded1052e))

#### Features

*   deprecate unqlite-sys and reexport `ffi` and `vars` module ([86747446](86747446))



<a name="v1.3.2"></a>
## v1.3.2 (2017-12-20)


#### Bug Fixes

*   implement `Send` and `Sync` for `Error` ([c3c93972](c3c93972))



<a name="v1.3.1"></a>
## v1.3.1 (2017-08-28)


#### Bug Fixes

*   fix error as Unique/Shared syntax changed ([1ce81552](1ce81552), closes [#5](5))



<a name="v1.3.0"></a>
## v1.3.0 (2017-06-28)


#### Features

* **api:**  document store interface (json via jx9) ([276eb35c](276eb35c))



<a name="v1.2.2"></a>
## v1.2.2 (2017-05-24)


#### Bug Fixes

*   fix in case of Shared.as_mut_ptr deprecation ([5a90c035](5a90c035))



<a name="v1.2.1"></a>
## v1.2.1 (2016-05-25)


#### Bug Fixes

* **cursor:**  fix delete error when reach the last ([9664f93d](9664f93d))



<a name="v1.2.0"></a>
## v1.2.0 (2016-05-13)


#### Features

* **UnQLite:**  use official name as struct name ([65b647e0](65b647e0))

#### Bug Fixes

* **clippy:**  fix some clippy warnings ([c877df4a](c877df4a))



<a name="v1.1.1"></a>
## v1.1.1 (2016-05-13)


#### Bug Fixes

* **unqlite-sys:**  update dependency to v1.0.0 ([9a0a05e2](9a0a05e2))



<a name="v1.1.0"></a>
## v1.1.0 (2016-05-12)


#### Features

* **cursor:**  add key_callback/value_callback method to `Entry` ([817ac5da](817ac5da))

#### Bug Fixes

*   set OpenMode as private ([d5270acd](d5270acd))
* **CI:**  fix unstable bug in travis-cargo ([be2c016d](be2c016d))



<a name="v1.0.0"></a>
## v1.0.0 (2016-05-12)






src/vars.rs
/* automatically generated by rust-bindgen */

pub const UNQLITE_VERSION: &'static [u8; 6usize] = b"1.1.8\0";
pub const UNQLITE_VERSION_NUMBER: ::std::os::raw::c_int = 1001008;
pub const UNQLITE_SIG: &'static [u8; 14usize] = b"unqlite/1.1.8\0";
pub const UNQLITE_IDENT: &'static [u8; 49usize] =
    b"unqlite:b172a1e2c3f62fb35c8e1fb2795121f82356cad6\0";
pub const UNQLITE_COPYRIGHT : & 'static [ u8 ; 108usize ] = b"Copyright (C) Symisc Systems, S.U.A.R.L [Mrad Chems Eddine <chm@symisc.net>] 2012-2018, http://unqlite.org/\0" ;
pub const SXRET_OK: ::std::os::raw::c_int = 0;
pub const SXERR_MEM: ::std::os::raw::c_int = -1;
pub const SXERR_IO: ::std::os::raw::c_int = -2;
pub const SXERR_EMPTY: ::std::os::raw::c_int = -3;
pub const SXERR_LOCKED: ::std::os::raw::c_int = -4;
pub const SXERR_ORANGE: ::std::os::raw::c_int = -5;
pub const SXERR_NOTFOUND: ::std::os::raw::c_int = -6;
pub const SXERR_LIMIT: ::std::os::raw::c_int = -7;
pub const SXERR_MORE: ::std::os::raw::c_int = -8;
pub const SXERR_INVALID: ::std::os::raw::c_int = -9;
pub const SXERR_ABORT: ::std::os::raw::c_int = -10;
pub const SXERR_EXISTS: ::std::os::raw::c_int = -11;
pub const SXERR_SYNTAX: ::std::os::raw::c_int = -12;
pub const SXERR_UNKNOWN: ::std::os::raw::c_int = -13;
pub const SXERR_BUSY: ::std::os::raw::c_int = -14;
pub const SXERR_OVERFLOW: ::std::os::raw::c_int = -15;
pub const SXERR_WILLBLOCK: ::std::os::raw::c_int = -16;
pub const SXERR_NOTIMPLEMENTED: ::std::os::raw::c_int = -17;
pub const SXERR_EOF: ::std::os::raw::c_int = -18;
pub const SXERR_PERM: ::std::os::raw::c_int = -19;
pub const SXERR_NOOP: ::std::os::raw::c_int = -20;
pub const SXERR_FORMAT: ::std::os::raw::c_int = -21;
pub const SXERR_NEXT: ::std::os::raw::c_int = -22;
pub const SXERR_OS: ::std::os::raw::c_int = -23;
pub const SXERR_CORRUPT: ::std::os::raw::c_int = -24;
pub const SXERR_CONTINUE: ::std::os::raw::c_int = -25;
pub const SXERR_NOMATCH: ::std::os::raw::c_int = -26;
pub const SXERR_RESET: ::std::os::raw::c_int = -27;
pub const SXERR_DONE: ::std::os::raw::c_int = -28;
pub const SXERR_SHORT: ::std::os::raw::c_int = -29;
pub const SXERR_PATH: ::std::os::raw::c_int = -30;
pub const SXERR_TIMEOUT: ::std::os::raw::c_int = -31;
pub const SXERR_BIG: ::std::os::raw::c_int = -32;
pub const SXERR_RETRY: ::std::os::raw::c_int = -33;
pub const SXERR_IGNORE: ::std::os::raw::c_int = -63;
pub const UNQLITE_OK: ::std::os::raw::c_int = 0;
pub const UNQLITE_NOMEM: ::std::os::raw::c_int = -1;
pub const UNQLITE_ABORT: ::std::os::raw::c_int = -10;
pub const UNQLITE_IOERR: ::std::os::raw::c_int = -2;
pub const UNQLITE_CORRUPT: ::std::os::raw::c_int = -24;
pub const UNQLITE_LOCKED: ::std::os::raw::c_int = -4;
pub const UNQLITE_BUSY: ::std::os::raw::c_int = -14;
pub const UNQLITE_DONE: ::std::os::raw::c_int = -28;
pub const UNQLITE_PERM: ::std::os::raw::c_int = -19;
pub const UNQLITE_NOTIMPLEMENTED: ::std::os::raw::c_int = -17;
pub const UNQLITE_NOTFOUND: ::std::os::raw::c_int = -6;
pub const UNQLITE_NOOP: ::std::os::raw::c_int = -20;
pub const UNQLITE_INVALID: ::std::os::raw::c_int = -9;
pub const UNQLITE_EOF: ::std::os::raw::c_int = -18;
pub const UNQLITE_UNKNOWN: ::std::os::raw::c_int = -13;
pub const UNQLITE_LIMIT: ::std::os::raw::c_int = -7;
pub const UNQLITE_EXISTS: ::std::os::raw::c_int = -11;
pub const UNQLITE_EMPTY: ::std::os::raw::c_int = -3;
pub const UNQLITE_COMPILE_ERR: ::std::os::raw::c_int = -70;
pub const UNQLITE_VM_ERR: ::std::os::raw::c_int = -71;
pub const UNQLITE_FULL: ::std::os::raw::c_int = -73;
pub const UNQLITE_CANTOPEN: ::std::os::raw::c_int = -74;
pub const UNQLITE_READ_ONLY: ::std::os::raw::c_int = -75;
pub const UNQLITE_LOCKERR: ::std::os::raw::c_int = -76;
pub const UNQLITE_CONFIG_JX9_ERR_LOG: ::std::os::raw::c_int = 1;
pub const UNQLITE_CONFIG_MAX_PAGE_CACHE: ::std::os::raw::c_int = 2;
pub const UNQLITE_CONFIG_ERR_LOG: ::std::os::raw::c_int = 3;
pub const UNQLITE_CONFIG_KV_ENGINE: ::std::os::raw::c_int = 4;
pub const UNQLITE_CONFIG_DISABLE_AUTO_COMMIT: ::std::os::raw::c_int = 5;
pub const UNQLITE_CONFIG_GET_KV_NAME: ::std::os::raw::c_int = 6;
pub const UNQLITE_VM_CONFIG_OUTPUT: ::std::os::raw::c_int = 1;
pub const UNQLITE_VM_CONFIG_IMPORT_PATH: ::std::os::raw::c_int = 2;
pub const UNQLITE_VM_CONFIG_ERR_REPORT: ::std::os::raw::c_int = 3;
pub const UNQLITE_VM_CONFIG_RECURSION_DEPTH: ::std::os::raw::c_int = 4;
pub const UNQLITE_VM_OUTPUT_LENGTH: ::std::os::raw::c_int = 5;
pub const UNQLITE_VM_CONFIG_CREATE_VAR: ::std::os::raw::c_int = 6;
pub const UNQLITE_VM_CONFIG_HTTP_REQUEST: ::std::os::raw::c_int = 7;
pub const UNQLITE_VM_CONFIG_SERVER_ATTR: ::std::os::raw::c_int = 8;
pub const UNQLITE_VM_CONFIG_ENV_ATTR: ::std::os::raw::c_int = 9;
pub const UNQLITE_VM_CONFIG_EXEC_VALUE: ::std::os::raw::c_int = 10;
pub const UNQLITE_VM_CONFIG_IO_STREAM: ::std::os::raw::c_int = 11;
pub const UNQLITE_VM_CONFIG_ARGV_ENTRY: ::std::os::raw::c_int = 12;
pub const UNQLITE_VM_CONFIG_EXTRACT_OUTPUT: ::std::os::raw::c_int = 13;
pub const UNQLITE_KV_CONFIG_HASH_FUNC: ::std::os::raw::c_int = 1;
pub const UNQLITE_KV_CONFIG_CMP_FUNC: ::std::os::raw::c_int = 2;
pub const UNQLITE_LIB_CONFIG_USER_MALLOC: ::std::os::raw::c_int = 1;
pub const UNQLITE_LIB_CONFIG_MEM_ERR_CALLBACK: ::std::os::raw::c_int = 2;
pub const UNQLITE_LIB_CONFIG_USER_MUTEX: ::std::os::raw::c_int = 3;
pub const UNQLITE_LIB_CONFIG_THREAD_LEVEL_SINGLE: ::std::os::raw::c_int = 4;
pub const UNQLITE_LIB_CONFIG_THREAD_LEVEL_MULTI: ::std::os::raw::c_int = 5;
pub const UNQLITE_LIB_CONFIG_VFS: ::std::os::raw::c_int = 6;
pub const UNQLITE_LIB_CONFIG_STORAGE_ENGINE: ::std::os::raw::c_int = 7;
pub const UNQLITE_LIB_CONFIG_PAGE_SIZE: ::std::os::raw::c_int = 8;
pub const UNQLITE_OPEN_READONLY: ::std::os::raw::c_uint = 1;
pub const UNQLITE_OPEN_READWRITE: ::std::os::raw::c_uint = 2;
pub const UNQLITE_OPEN_CREATE: ::std::os::raw::c_uint = 4;
pub const UNQLITE_OPEN_EXCLUSIVE: ::std::os::raw::c_uint = 8;
pub const UNQLITE_OPEN_TEMP_DB: ::std::os::raw::c_uint = 16;
pub const UNQLITE_OPEN_NOMUTEX: ::std::os::raw::c_uint = 32;
pub const UNQLITE_OPEN_OMIT_JOURNALING: ::std::os::raw::c_uint = 64;
pub const UNQLITE_OPEN_IN_MEMORY: ::std::os::raw::c_uint = 128;
pub const UNQLITE_OPEN_MMAP: ::std::os::raw::c_uint = 256;
pub const UNQLITE_SYNC_NORMAL: ::std::os::raw::c_uint = 2;
pub const UNQLITE_SYNC_FULL: ::std::os::raw::c_uint = 3;
pub const UNQLITE_SYNC_DATAONLY: ::std::os::raw::c_uint = 16;
pub const UNQLITE_LOCK_NONE: ::std::os::raw::c_uint = 0;
pub const UNQLITE_LOCK_SHARED: ::std::os::raw::c_uint = 1;
pub const UNQLITE_LOCK_RESERVED: ::std::os::raw::c_uint = 2;
pub const UNQLITE_LOCK_PENDING: ::std::os::raw::c_uint = 3;
pub const UNQLITE_LOCK_EXCLUSIVE: ::std::os::raw::c_uint = 4;
pub const UNQLITE_ACCESS_EXISTS: ::std::os::raw::c_uint = 0;
pub const UNQLITE_ACCESS_READWRITE: ::std::os::raw::c_uint = 1;
pub const UNQLITE_ACCESS_READ: ::std::os::raw::c_uint = 2;
pub const UNQLITE_CURSOR_MATCH_EXACT: ::std::os::raw::c_uint = 1;
pub const UNQLITE_CURSOR_MATCH_LE: ::std::os::raw::c_uint = 2;
pub const UNQLITE_CURSOR_MATCH_GE: ::std::os::raw::c_uint = 3;
pub const UNQLITE_JOURNAL_FILE_SUFFIX: &'static [u8; 17usize] = b"_unqlite_journal\0";
pub const UNQLITE_CTX_ERR: ::std::os::raw::c_uint = 1;
pub const UNQLITE_CTX_WARNING: ::std::os::raw::c_uint = 2;
pub const UNQLITE_CTX_NOTICE: ::std::os::raw::c_uint = 3;
